### PR TITLE
ATO 207 run form validation  on form activation

### LINKED ISSUE #11199: required_slots function in form is not called
h3\.   Rasa Open Source version

3.1.0

h3\.   Rasa SDK version

3.1.2

h3\.   Python version

3.7

h3\.   What operating system are you using?

Windows

h3\.   What happened?

I've created a form with form validation. Form doesn't call my required_slots method in spite of I've created a custom required_slots function. This is my code:

It calls firstly action_ask_totalizator_block_form_card_number action. But I expect firstly to go to my overriden required_slots function. (I tested on other forms and this works well)

    class ValidateTotalizatorBlockForm(FormValidationAction):

        def name(self) -> Text:
            """Unique identifier of the form. """
            return "validate_totalizator_block_form"

        async def required_slots(
                self,
                domain_slots: List[Text],
                dispatcher: "CollectingDispatcher",
                tracker: "Tracker",
                domain: "DomainDict",
        ) -> Optional[List[Text]]:
            token = get_metadata_parameters(tracker, key="token")
            if not token:
                return []

            number = tracker.get_slot("number")
            card_number = tracker.get_slot("card_number")
            otp = tracker.get_slot("otp")
            accept = tracker.get_slot("accept")

            logger.info(f"number {number}")
            logger.info(f"card_number {card_number}")
            logger.info(f"otp {otp}")
            logger.info(f"accept {accept}")

            if card_number and accept == "no":
                return []

            return ["card_number", "accept", "otp"]

    domain.yml
      - action_ask_totalizator_block_form_card_number
      - action_ask_totalizator_block_form_otp
      - action_submit_totalizator_block_form
      - validate_totalizator_block_form

      totalizator_block_form:
        required_slots:
          - card_number
          - accept
          - otp

    rules:
    - rule: Activate totalizator_block_form
      steps:
      - intent: totalizator.block
      - action: totalizator_block_form
      - active_loop: totalizator_block_form # Keep requesting form slots until they're filled in

    - rule: Submit totalizator_block_form
      condition: # Condition that form is active
      - active_loop: totalizator_block_form
      steps:
      - action: totalizator_block_form
      - active_loop: null # Stop the form
      - slot_was_set:
        - requested_slot: null
      - action: action_submit_totalizator_block_form

h3\.   Command / Request

When I trigger `totalizator.block` intent, it goes to my `action_ask_totalizator_block_form_card_number`. I want it to go to required_slots function firstly.

h3\.   Relevant log output

<pre>Traceback (most recent call last):
  File "handle_request", line 83, in handle_request
    )
  File "C:\Users\igeladze\AppData\Local\Continuum\anaconda3\envs\rasa3\lib\site-packages\rasa_sdk\endpoint.py", line 104, in webhook
    result = await executor.run(action_call)
  File "C:\Users\igeladze\AppData\Local\Continuum\anaconda3\envs\rasa3\lib\site-packages\rasa_sdk\executor.py", line 398, in run
    action(dispatcher, tracker, domain)
  File "C:\Users\igeladze\AppData\Local\Continuum\anaconda3\envs\rasa3\lib\site-packages\rasa_sdk\utils.py", line 230, in call_potential_coroutine
    return await coroutine_or_return_value
  File "C:\Users\igeladze\Desktop\BOG_Projects\Chatbot3\actions\form\totalizator_block_form.py", line 287, in run
    message, buttons, card_number_and_card_id_map = self.get_card_list(tracker)
  File "C:\Users\igeladze\Desktop\BOG_Projects\Chatbot3\actions\form\totalizator_block_form.py", line 239, in get_card_list
    channel = chat_rb_channel_mapping(channel)
  File "C:\Users\igeladze\Desktop\BOG_Projects\Chatbot3\common\general.py", line 65, in chat_rb_channel_mapping
    assert chat_channel in channels_mapping</pre>

This happens because it calls action_ask_totalizator_block_form_card_number function and not my required_slots function. In my required_slots function, I return [] if user is not logged in. Then it should go to submit and this error will not happen. My question is why it doesnt go to my required_slots function firstly.

### PR DESCRIPTION
**Pull request by melindaloubser1**
**Proposed changes**:
- Revert change here https://github.com/RasaHQ/rasa/pull/10295/files#r807244104 that removed running the form validation action on activation of the form (before the loop is active).
- This would make #10913 obsolete since the 2.x mechanism for achieving the same thing would be returned.
- Closes #11199 
- Resolves the issue where any `extract_<slot_name>` method in a FormValidationAction subclass only runs after a required_slot in the form has been set (i.e. after activation).

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: tests/core/actions/test_forms.py [hunk 1]
```diff
@@ -34,38 +35,69 @@
 from rasa.shared.core.trackers import DialogueStateTracker
 from rasa.utils.endpoints import EndpointConfig
 
+ACTION_SERVER_URL = "http://my-action-server:5055/webhook"
+
 
 async def test_activate():
```
Comment (ancalita) [inline tests/core/actions/test_forms.py:41]: I'd actually recommend keeping this unit test unchanged, and then writing a new test with the custom slot as you've written here.

File: tests/core/actions/test_forms.py [hunk 2]
```diff
@@ -761,6 +786,10 @@ async def test_validate_slots_on_activation_with_other_action_after_user_utteran
 
         form_action = FormAction(form_name, action_server)
 
+        mocked.post(
```
Comment (ancalita) [inline tests/core/actions/test_forms.py:789]: Why are these lines required now? 🤔

File: tests/core/actions/test_forms.py [hunk 3]
```diff
@@ -5,6 +5,7 @@
 import pytest
 from _pytest.monkeypatch import MonkeyPatch
 from aioresponses import aioresponses
+from questionary import form
```
Comment (ancalita) [inline tests/core/actions/test_forms.py:8]: Why do you need this import, it seems to cause a lot of code quality fails?

**Review by ancalita — COMMENTED** [review]
Looks good, an additional q I have is to double-check there was no need to add this check back into `_get_slot_extractions` for loop which was removed in PR 10295?
```
if not tracker.active_loop:
  break
```

File: tests/core/actions/test_forms.py (hunk 2 above)
Comment (melindaloubser1) [reply tests/core/actions/test_forms.py:789]: Because the validation action now runs during the subsequent lines, it throws a "can't reach action server" exception without these lines.

File: tests/core/actions/test_forms.py (hunk 3 above)
Comment (melindaloubser1) [reply tests/core/actions/test_forms.py:8]: entirely accidental, probably an auto import by my IDE

File: tests/core/actions/test_forms.py (hunk 2 above)
Comment (ancalita) [reply tests/core/actions/test_forms.py:789]: Oh I see, because we have to mimic running the validation action at form initialization 👍🏻

File: tests/core/actions/test_forms.py [hunk 4]
```diff
@@ -66,6 +68,68 @@ async def test_activate():
...
+        )
+    assert events[:-1] == [
+        ActiveLoop(form_name),
+        SlotSet(
+            slot_set_by_remote_custom_extraction_method,
+            slot_value_set_by_remote_custom_extraction_method,
+        ),
+        SlotSet(REQUESTED_SLOT, domain_required_slot_name),
+    ]
     assert isinstance(events[-1], BotUttered)
```
Comment (ancalita) [inline tests/core/actions/test_forms.py:133]: Could we have this assertion back into `test_activate` too? 😁

**Review by ancalita — APPROVED** [review]
Whoop whoop 🎉 
Left one tiny comment to make sure we have that wondering assert back into `test_activate` too, otherwise good to merge.

**Review by losterloh — APPROVED** [review]
🚀🚀🚀

**Comment by XiaofeiQian** [comment]
Can we merge and release this soon? thanks.
